File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
import re

arr = [
    ["+", "-", "*"],
    ["+", "*", "-"],
    ["-", "+", "*"],
    ["-", "*", "+"],
    ["*", "+", "-"],
    ["*", "-", "*"],
]

def calculate(left, right, op):
    if op == "+":
        return left + right
    elif op == "-":
        return left - right
    else:
        return left * right

def solve(ops, tokens):

    for op in ops:
        
        i = 0
        while i < len(tokens):

            if tokens[i] == op:
                left = int(tokens[i - 1])
                right = int(tokens[i + 1])
                result = calculate(left, right, op)

                tokens[i - 1:i + 2] = [str(result)]
                i -= 2

            i += 1

    return abs(int(tokens[0]))

def solution(expression):
    
    tokens = re.split(r'([+\-*])', expression)

    ans = 0

    for ops in arr:
        num = solve(ops, tokens[:])
        ans = max(ans, num)

    return ans

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.*;

class Solution {
    
    static String[][] arr = {
            "+-*".split(""),
            "+*-".split(""),
            "-+*".split(""),
            "-*+".split(""),
            "*+-".split(""),
            "*-+".split(""),
    };
    
    public long solution(String expression) {
        StringTokenizer st = new StringTokenizer(expression, "+-*", true);

        List<String> tokens = new ArrayList<>();

        while (st.hasMoreTokens()) {
            tokens.add(st.nextToken());
        }

        long ans = 0;

        for (String[] ops : arr) {

            long num = solve(ops, new LinkedList<>(tokens));
            ans = Math.max(ans, num);
        }

        return ans;
    }
    
    public long solve(String[] ops, List<String> tokens) {

        for (String op : ops) {
            
        ListIterator<String> iter = tokens.listIterator();

            while (iter.hasNext()) {

                String cur = iter.next();

                if (cur.equals(op)) {

                    iter.previous();
                    long left = Long.parseLong(iter.previous());
                    
                    iter.next();
                    iter.next();
                    long right = Long.parseLong(iter.next());

                    long result = calculate(left, right, cur);

                    for (int i = 0; i < 3; i++) {
                        iter.previous();
                        iter.remove();
                    }

                    iter.add(String.valueOf(result));
                }
            }
        }
        
        return Math.abs(Long.parseLong(tokens.get(0)));
    }
    
    public long calculate(long left, long right, String op) {
        switch (op) {
            case "+":
                return left + right;
            case "-":
                return left - right;
            case "*":
                return left * right;
        }
        return 0;
    }
}